### PR TITLE
[wasm64] Fix a lot of the remaining browser tests

### PR DESCRIPTION
**Pull request by sbc100**
This also inverts the testing logic such that we run all browser tests in wasm64 mode.  Tests that don't currently pass are annotated with @no_wasm64.

### CODE REVIEW

**Comment by sbc100** [comment]
I think this change is now small enough to land, having split ~10 changes from it.

PTAL.

File: system/lib/compiler-rt/lib/sanitizer_common/sanitizer_emscripten.cpp [hunk 1]
```diff
@@ -127,6 +127,9 @@ u64 MonotonicNanoTime() {
   return (u64)ts.tv_sec * (1000ULL * 1000 * 1000) + ts.tv_nsec;
 }
 
+// FIXME implement on this platform.
+void GetMemoryProfile(fill_profile_f cb, uptr *stats) {}
```
Comment (kripken) [inline system/lib/compiler-rt/lib/sanitizer_common/sanitizer_emscripten.cpp:131]: How does this help get a test to pass? Is this wasm64-specific?

File: test/test_mem_growth.c [hunk 2]
```diff
@@ -5,7 +5,11 @@
 #include <emscripten/html5.h>
 #include <emscripten/heap.h>
 
+#ifdef __wasm64__
+#define EXPECTED_FINAL_HEAP_SIZE (4*1024*1024*1024ll)
```
Comment (kripken) [inline test/test_mem_growth.c:9]: Interesting, why is wasm64 capable of allocating 64K more here?

File: test/wasmfs/wasmfs_fetch.js [hunk 3]
```diff
@@ -1,5 +1,7 @@
 addToLibrary({
   getUrlOrigin: function (ptr, len) {
+    ptr = Number(ptr);
+    ptr = Number(len);
```
Comment (kripken) [inline test/wasmfs/wasmfs_fetch.js:4]: Should this not be closer to where the pointer arrives from wasm, and use the usual non-manual macros to convert as needed?

File: test/test_browser.py [hunk 4]
```diff
@@ -2678,6 +2723,11 @@ def test_doublestart_bug(self):
   })
   @requires_threads
   def test_html5_core(self, opts):
+    if self.is_wasm64():
+      if '-sMIN_CHROME_VERSION=0' in opts:
+        self.skipTest('wasm64 doesnt support older browsers')
```
Comment (kripken) [inline test/test_browser.py:2728]:
```suggestion
        self.skipTest('wasm64 does not support older browsers')
```

File: test/test_browser.py [hunk 5]
```diff
@@ -2678,6 +2723,11 @@ def test_doublestart_bug(self):
...
+      if '-sPROXY_TO_PTHREAD' in opts:
+        self.skipTest('_emscripten_set_keypress_callback_on_thread broken')
```
Comment (kripken) [inline test/test_browser.py:2730]: Is this related to wasm64?

File: test/test_browser.py [hunk 6]
```diff
@@ -3910,6 +4014,7 @@ def test_pthread_gcc_atomic_fetch_and_op(self):
         self.btest_exit('pthread/test_pthread_gcc_atomic_fetch_and_op.cpp', args=args + ['-sINITIAL_MEMORY=64MB', '-pthread', '-sPTHREAD_POOL_SIZE=8'])
 
   # 64 bit version of the above test.
+  @no_wasm64("XXX")
```
Comment (kripken) [inline test/test_browser.py:4017]: No reason?

File: test/test_browser.py [hunk 7]
```diff
@@ -3926,6 +4031,7 @@ def test_pthread_gcc_atomic_op_and_fetch(self):
     self.btest_exit('pthread/test_pthread_gcc_atomic_op_and_fetch.cpp', args=['-sINITIAL_MEMORY=64MB', '-O3', '-pthread', '-sPTHREAD_POOL_SIZE=8'])
 
   # 64 bit version of the above test.
+  @no_wasm64("XXX")
```
Comment (kripken) [inline test/test_browser.py:4034]: ditto

**Review by kripken — COMMENTED** [review]
lgtm % comments

File: test/test_mem_growth.c (hunk 2 above)
Comment (sbc100) [reply test/test_mem_growth.c:9]: According to library.js:

```
    // Stay one Wasm page short of 4GB: while e.g. Chrome is able to allocate       
    // full 4GB Wasm memories, the size will wrap back to 0 bytes in Wasm side   
    // for any code that deals with heap sizes, which would require special         
    // casing all heap size related code to treat 0 specially.                   
    {{{ Math.min(MAXIMUM_MEMORY, FOUR_GB - WASM_PAGE_SIZE) }}},   
```

Comment (kripken) [reply test/test_mem_growth.c:9]: I see, thanks. Maybe refer to that from here in the test?

File: test/test_browser.py (hunk 5 above)
Comment (sbc100) [reply test/test_browser.py:2730]: Yes _emscripten_set_keypress_callback_on_thread doesn't work on wasm64.. I'll update the comment

File: test/wasmfs/wasmfs_fetch.js (hunk 3 above)
Comment (sbc100) [reply test/wasmfs/wasmfs_fetch.js:4]: You are right.. I'm not sure why it didn't just use `__sig` here.

File: test/test_browser.py (hunk 6 above)
Comment (sbc100) [reply test/test_browser.py]: Reverted

File: test/test_browser.py (hunk 7 above)
Comment (sbc100) [reply test/test_browser.py]: Reverted

File: test/test_mem_growth.c (hunk 2 above)
Comment (sbc100) [reply test/test_mem_growth.c]: Done

File: system/lib/compiler-rt/lib/sanitizer_common/sanitizer_emscripten.cpp (hunk 1 above)
Comment (sbc100) [reply system/lib/compiler-rt/lib/sanitizer_common/sanitizer_emscripten.cpp]: Reverted

File: test/test_mem_growth.c [hunk 8]
```diff
@@ -5,7 +5,13 @@
 #include <emscripten/html5.h>
 #include <emscripten/heap.h>
 
-#define EXPECTED_FINAL_HEAP_SIZE (4*1024*1024*1024ll - 65536)
+#ifdef __wasm64__
+#define MAX_HEAP (4*1024*1024*1024ll)
+#else
+// We don't allow the full 4Gb on wasm32 since that size would wrap
+// back to zero.  Se getHeapMax in library.js.
```
Comment (kripken) [inline test/test_mem_growth.c:12]:
```suggestion
// back to zero.  See getHeapMax in library.js.
```

**Review by kripken — APPROVED** [review]
